exp/controllers/azuremachinepool_controller.go

```diff
@@ -0,0 +1,587 @@
+	ampSpec := s.machinePoolScope.AzureMachinePool.Spec
+	var replicas int64
+	if s.machinePoolScope.MachinePool.Spec.Replicas != nil {
+		replicas = int64(to.Int32(s.machinePoolScope.MachinePool.Spec.Replicas))
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(ampSpec.Template.SSHPublicKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to base64 decode ssh public key")
+	}
+
+	image, err := getVMImage(s.machinePoolScope)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get VM image")
+	}
+
+	bootstrapData, err := s.machinePoolScope.GetBootstrapData()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve bootstrap data")
+	}
+
+	vmssSpec := &scalesets.Spec{
+		Name:       s.machinePoolScope.Name(),
+		Sku:        ampSpec.Template.VMSize,
+		Capacity:   replicas,
+		SSHKeyData: string(decoded),
+		Image:      image,
+		OSDisk:     ampSpec.Template.OSDisk,
+		CustomData: bootstrapData,
+	}
+
+	err = s.virtualMachinesScaleSetSvc.Reconcile(context.TODO(), vmssSpec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create or get machine")
+	}
+
+	newVMSS, err := s.virtualMachinesScaleSetSvc.Get(s.clusterScope.Context, vmssSpec)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get vm")
+	}
+
+	vmss, ok := newVMSS.(*infrav1exp.VMSS)
+	if !ok {
+		return nil, errors.New("returned incorrect vmss interface")
+	}
+	if vmss.State == "" {
+		return nil, errors.Errorf("vmss %s is nil provisioning state, reconcile", s.machinePoolScope.Name())
+	}
+
+	if vmss.State == infrav1.VMStateFailed {
+		// If VM failed provisioning, delete it so it can be recreated
+		err = s.virtualMachinesScaleSetSvc.Delete(s.clusterScope.Context, vmssSpec)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to delete machine pool")
+		}
+		return nil, errors.Errorf("vmss %s is deleted, retry creating in next reconcile", s.machinePoolScope.Name())
+	} else if vmss.State != infrav1.VMStateSucceeded {
+		return nil, errors.Errorf("vm %s is still in provisioningState %s, reconcile", s.machinePoolScope.Name(), vmss.State)
+	}
+
+	return vmss, nil
+}
+
+// Delete reconciles all the services in pre determined order
+func (s *azureMachinePoolService) Delete() error {
+	vmssSpec := &scalesets.Spec{
+		Name: s.machinePoolScope.Name(),
+	}
+
+	err := s.virtualMachinesScaleSetSvc.Delete(s.clusterScope.Context, vmssSpec)
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete machine pool")
+	}
+
+	return nil
+}
+
+// Get fetches a VMSS if it exists
+func (s *azureMachinePoolService) Get() (*infrav1exp.VMSS, error) {
+	vmssSpec := &scalesets.Spec{
+		Name: s.machinePoolScope.Name(),
+	}
+
+	vmss, err := s.virtualMachinesScaleSetSvc.Get(s.clusterScope.Context, vmssSpec)
+	if err != nil && !azure.ResourceNotFound(err) {
+		return nil, errors.Wrapf(err, "failed to fetch machine pool")
+	}
+
+	if err != nil && azure.ResourceNotFound(err) {
+		return nil, nil
```

Comment:
why are we returning a nil error here?

Reply:
I really didn't think it mattered at the next level since it was resource not found. At the next level up, we'd just have to check for the error again to see if we got a 404. What's the point?

Reply:
I usually expect a non-nil object if error is nil.